### PR TITLE
Alerting: Migrate dashboard alert permissions

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
It targets #33000
For the time being, migration triggers on environment variable e.g. `bash: $ UALERT_MIG=iDidBackup make run` (and expects the `alert_rule` table to be created -- currently it's created if the AlertNG service is enabled)

It migrates dashboard alert permissions by creating the necessary folders as follows:
- If alert's dashboard has permissions, it creates corresponding folder to match permissions (including default and inherited)
- If no dashboard permissions, and folder, it links to the new alert rule to the existing folder
- If no dashboard permissions and no folder (General/root) - it links to a new folder per organisation named like `_general_alerting_org_<orgId>__` with no perms to inherit default perms

the migrated alert rules organised by folders:
Screenshot 2021-04-28 at 5 03 37 PM

the created general folder `_general_alerting_org_1__` permissions:
Screenshot 2021-04-28 at 5 03 52 PM

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by kylebrandt** [comment]
Is the orgId needed in the folder name? I assumed you could have duplicate folder names across different orgs?

File: pkg/services/sqlstore/migrations/ualert/permissions.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/util"
+
+	"github.com/grafana/grafana/pkg/infra/metrics"
+	"github.com/grafana/grafana/pkg/models"
```
Comment (kylebrandt) [inline pkg/services/sqlstore/migrations/ualert/permissions.go:11]: Don't know that this it is safe to use these models / commands. As if the models change in the future, the migration will have changed.

**Comment by papagian** [comment]
> Is the orgId needed in the folder name? I assumed you could have duplicate folder names across different orgs?

asking about the general folder? There is a unique constrain on `org_id`, `folder_id`, `title`; `folder_id` (the parent folder) for folders is always 0 so we can duplicate folder names across different orgs.
I'm not happy with the naming of the new folders; I'm open to any suggestions.

**Comment by kylebrandt** [comment]
> asking about the general folder? 
Yup, that is the one I was thinking would not need the orgId suffix in the name

File: pkg/services/sqlstore/migrations/ualert/permissions.go (hunk 1 above)
Comment (papagian) [reply pkg/services/sqlstore/migrations/ualert/permissions.go:11]: It makes sense; I will change it

**Comment by kylebrandt** [comment]
Seems to work on my local instance :-)

File: pkg/services/sqlstore/migrations/ualert/permissions.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+)
+
+// getOrCreateGeneralFolder returns the general folder under the specific organisation
+// If the general folder does not exist it creates it.
+func (m *migration) getOrCreateGeneralFolder(orgID int64) (*models.Dashboard, error) {
```
Comment (torkelo) [inline pkg/services/sqlstore/migrations/ualert/permissions.go:16]: Any reason for creating the general folder? I do not think we have ever created this folder, but might be a good idea still as you cannot set permissions on general today as it does not actually exist

Comment (kylebrandt) [reply pkg/services/sqlstore/migrations/ualert/permissions.go:16]: I think the main reason we said was to lessen the chance that we would pollute the root with too many items during the migration.

**Comment by papagian** [comment]
@kylebrandt I have noticed in the above shared screenshot that `TestData - Always Pending` and `TestData - No data` states are not the correct (they should have been `Pending` and `No data` respectively).

However the transformation seems to be correct for the `TestData - Always Pending` alert:
Screenshot 2021-04-29 at 1 19 09 PM
Screenshot 2021-04-29 at 1 12 13 PM

and the `TestData - No data` alert:
Screenshot 2021-04-29 at 1 11 14 PM
Screenshot 2021-04-29 at 1 24 49 PM

**Review by kylebrandt — APPROVED** [review]

**Pull request merged**